File: akkaStreamsCustomStage/src/main/scala/com/lightbend/scala/modelServer/modelServer/ModelStage.scala
package com.lightbend.scala.modelServer.modelServer

import java.util.concurrent.TimeUnit

import akka.stream._
import akka.stream.stage._
import com.lightbend.model.winerecord.WineRecord
import com.lightbend.scala.modelServer.model.{Model, ModelToServeStats, ModelWithDescriptor, ServingResult}


class ModelStage extends GraphStageWithMaterializedValue[FlowShape[WineRecord, ServingResult], ModelStateStore] {

  val dataRecordIn = Inlet[WineRecord]("dataRecordIn")
  val scoringResultOut = Outlet[ServingResult]("scoringOut")

  override val shape: FlowShape[WineRecord, ServingResult] = FlowShape(dataRecordIn, scoringResultOut)

  class ModelLogic extends GraphStageLogicWithLogging(shape) {
    // state must be kept in the Logic instance, since it is created per stream materialization
    private var currentModel: Option[Model] = None
    private var newModel: Option[Model] = None
    var currentState: Option[ModelToServeStats] = None // exposed in materialized value
    private var newState: Option[ModelToServeStats] = None

    val setModelCB = getAsyncCallback[ModelWithDescriptor] { model =>
      println(s"Updated model: $model")
      newState = Some(ModelToServeStats(model.descriptor))
      newModel = Some(model.model)
    }

    setHandler(dataRecordIn, new InHandler {
      override def onPush(): Unit = {
        val record = grab(dataRecordIn)
        newModel.foreach { model =>
          // close current model first
          currentModel.foreach(_.cleanup())
          // Update model
          currentModel = Some(model)
          currentState = newState
          newModel = None
        }
        currentModel match {
          case Some(model) =>
            val start = System.nanoTime()
            val quality = model.score(record).asInstanceOf[Double]
            val duration = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start)
            currentState = currentState.map(_.incrementUsage(duration))
            push(scoringResultOut, ServingResult(quality, duration))

          case None =>
            push(scoringResultOut, ServingResult.noModel)
        }
      }
    })

    setHandler(scoringResultOut, new OutHandler {
      override def onPull(): Unit = {
        pull(dataRecordIn)
      }
    })
  }

  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, ModelStateStore) = {
    val logic = new ModelLogic

    // we materialize this value so whoever runs the stream can get the current serving info
    val modelStateStore = new ModelStateStore {
      override def getCurrentServingInfo: ModelToServeStats =
        logic.currentState.getOrElse(ModelToServeStats.empty)

      override def setModel(model: ModelWithDescriptor): Unit =
        logic.setModelCB.invoke(model)
    }
    (logic, modelStateStore)
  }
}

File: akkaStreamsCustomStage/src/main/java/com/lightbend/java/modelserver/modelserver/ModelStage.java
package com.lightbend.java.modelserver.modelserver;

import akka.japi.Pair;
import akka.japi.function.Procedure;
import akka.stream.Attributes;
import akka.stream.FlowShape;
import akka.stream.Inlet;
import akka.stream.Outlet;
import akka.stream.stage.*;
import com.lightbend.java.model.Model;
import com.lightbend.java.model.ModelServingInfo;
import com.lightbend.java.model.ModelWithDescriptor;
import com.lightbend.java.model.ServingResult;
import com.lightbend.model.Winerecord;

import java.util.Optional;

public class ModelStage extends AbstractGraphStageWithMaterializedValue<FlowShape<Winerecord.WineRecord, ServingResult>, ReadableModelStore> {

    Inlet<Winerecord.WineRecord> dataRecordIn = Inlet.create("dataRecordIn");
    Outlet<ServingResult> scoringResultOut = Outlet.create("scoringOut");
    private final FlowShape<Winerecord.WineRecord, ServingResult> shape = FlowShape.of(dataRecordIn, scoringResultOut);

    @Override
    public FlowShape<Winerecord.WineRecord, ServingResult> shape() { return shape; }

    private class ModelLogic extends GraphStageLogicWithLogging{
        // state must be kept in the Logic instance, since it is created per stream materialization
        Optional<Model> currentModel = Optional.empty();
        Optional<Model> newModel = Optional.empty();
        Optional<ModelServingInfo> currentServingInfo = Optional.empty();
        Optional<ModelServingInfo> newServingInfo = Optional.empty();

        public ModelLogic(FlowShape<Winerecord.WineRecord, ServingResult> shape){
            super(shape);
            setHandler(dataRecordIn, new AbstractInHandler() {
                @Override
                public void onPush() throws Exception {
                    Winerecord.WineRecord dataRecord = grab(dataRecordIn);
                    if (newModel.isPresent()) {
                        // update the model
                        if (currentModel.isPresent())
                            currentModel.get().cleanup();
                        currentModel = newModel;
                        currentServingInfo = newServingInfo;
                        newModel = Optional.empty();
                    }

                    // Actually score
                    if (currentModel.isPresent()) {
                        // Score the model
                        long start = System.currentTimeMillis();
                        double quality = (double) currentModel.get().score(dataRecord);
                        long duration = System.currentTimeMillis() - start;
                        currentServingInfo.get().update(duration);
//                        System.out.println("Calculated quality - " + quality + " in " + duration + "ms");
                        push(scoringResultOut, new ServingResult(quality, duration));
                    } else {
                        // No model currently
//                        System.out.println("No model available - skipping");
                        push(scoringResultOut, ServingResult.noModel);
                    }
                }
            });
            setHandler(scoringResultOut, new AbstractOutHandler() {
                @Override
                public void onPull() throws Exception {
                    pull(dataRecordIn);
                }
            });
        }

        AsyncCallback<ModelWithDescriptor> callback = createAsyncCallback(new Procedure<ModelWithDescriptor>() {
            @Override
            public void apply(ModelWithDescriptor model) throws Exception {
                System.out.println("Updated model: " + model);
                newModel = Optional.of(model.getModel());
                newServingInfo = Optional.of(new ModelServingInfo(model.getDescriptor().getName(),
                        model.getDescriptor().getDescription(), System.currentTimeMillis()));
            }
        });
    }

    @Override
    public Pair<GraphStageLogic, ReadableModelStore> createLogicAndMaterializedValuePair(Attributes inheritedAttributes) throws Exception {
        ModelLogic logic = new ModelLogic(shape);

        // we materialize this value so whoever runs the stream can get the current serving info
        ReadableModelStore modelStateStore = new ReadableModelStore() {
            @Override
            public ModelServingInfo getCurrentServingInfo() {
                return logic.currentServingInfo.orElse(ModelServingInfo.empty);
            }

            @Override
            public void setModel(ModelWithDescriptor model) {
                logic.callback.invoke(model);
            }
        };
        return new Pair<>(logic, modelStateStore);
    }
}
